### PR TITLE
Adds missing Data.Num.Linear.* instances for Word, Integer, Natural, Float, Word8/16/32/64 Int8/16/32/64

### PR DESCRIPTION
**Pull request by Qqwy**
This PR fixes two (related) issues (#403 and #466):

- Movable (and derived `Consumable`/`Dupable`) instances are added for `Integer` and `Natural`. This is safe since these types are represented as strict sum types whose elements are unlifed unboxed primitives so there is no way laziness might break linearity in there. (See also the [last part of the discussion in #403](https://github.com/tweag/linear-base/issues/403#issuecomment-1706679338))
- Instances for the typeclasses in `Data.Num.linear` (`Additive`/`AddIdentity`/`AdditiveGroup`/`Multiplicative`/`MultIdentity`/`Semiring`/`Ring`/`Num`) for the general-purpose number types that exist in the Prelude/`base` which were still missing them:
  - `Word`
  - `Float`
  - `Integer`
  - `Natural`
  - `Word8`/`Word16`/`Word32`/`Word64`
  - `Int8`/`Int16`/`Int32`/`Int64`

Still missing are:
- Instances for `Ratio a` and `Complex a`.
- Instances for `Foreign.C.Types` like `CInt`/`CFloat` etc.
- Instances for `Foreign.Ptr` types like `IntPtr`, `WordPtr`.

Those could be added just as easily but because I was not 100% sure whether adding linear instances for them would be (morally) correct or have unintended consequences that might break linearity in ineffable ways, I left them alone for now.

### CODE REVIEW

**Comment by Qqwy** [comment]
The `Movable` instances introduces in this PR pattern matches down to the internals of `Integer` and `Natural` just like was done for `Word`/`Int` etc. before. Doing so is only possible by pulling in the `ghc-bignum` package. As far as I can tell `base` itself depends on it so I believe adding this 'extra' dependency does not bloat the library in any way.

File: src/Data/Num/Linear.hs
Comment (Qqwy) [inline src/Data/Num/Linear.hs:214]: I took the liberty of re-ordering these by instance rather than by class because I believe it's easier on the eyes. Ormolu is disabled in this part of the code because it would add a newline between each `deriving` line. I hope you agree. If not, we can always change it back.

File: src/Data/Num/Linear.hs
Comment (aspiwack) [inline src/Data/Num/Linear.hs:353]: Please add a newline to please git
```suggestion
{- ORMOLU_ENABLE -}

```

File: src/Data/Num/Linear.hs
Comment (aspiwack) [reply src/Data/Num/Linear.hs:214]: Disabling Ormolu is a little unorthodox, but I must admit that in this case, it looks like it's worth it.

**Review by aspiwack — COMMENTED** [review]
This all looks good. Thank you very much. I noticed that the CI hadn't ran yet. So I triggered it, and I'll (auto)merge when green.

**Comment by aspiwack** [comment]
Actually, can I, instead, ask you to add the newline to the appropriate commit to avoid a silly commit? Sorry about the rather trivial request.

**Comment by Qqwy** [comment]
Ormolu/git should be happy now; of course the CI only runs after being approved once more.

**Comment by aspiwack** [comment]
I'm guessing Github will keep requiring an approval for CI until you have at least one commit on master. I relaunched it.

**Review by aspiwack — APPROVED** [review]

**Pull request merged**
